### ISSUE TITLE
Pybind11 "aligned_alloc" Error for g++ 8.2.0 with C++17

### ISSUE THREAD

**Issue by phil-zxx** [pybind/pybind11#1827]
Lets consider the following setup:

```cmake
## CMakeLists.txt

cmake_minimum_required(VERSION 3.13)
project(pybind_test)

set(CMAKE_CXX_STANDARD 17)

add_library(${PROJECT_NAME} SHARED main.cpp)
target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/pybind11/include)
target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/python_linux/include)
```

```cpp
// main.cpp

#include <pybind11/pybind11.h>

PYBIND11_MODULE(pybind_test, m) { }
```

Using `gcc-8.2.0/bin/g++` I get the following error:
```cpp
====================[ Build | pybind_test | Debug ]=============================
/app/clion-2019.1/bin/cmake/linux/bin/cmake --build /home/pybind_test/cmake-build-debug --target pybind_test -- -j 36
[ 50%] Building CXX object CMakeFiles/pybind_test.dir/main.cpp.o
In file included from /app/x86_64-gcc8.2.x/gcc-8.2.0/include/c++/8.2.0/stdlib.h:36,
                 from /home/pybind_test/python_linux/include/Python.h:34,
                 from /home/pybind_test/pybind11/include/pybind11/detail/common.h:112,
                 from /home/pybind_test/pybind11/include/pybind11/pytypes.h:12,
                 from /home/pybind_test/pybind11/include/pybind11/cast.h:13,
                 from /home/pybind_test/pybind11/include/pybind11/attr.h:13,
                 from /home/pybind_test/pybind11/include/pybind11/pybind11.h:44,
                 from /home/pybind_test/main.cpp:1:
/app/x86_64-gcc8.2.x/gcc-8.2.0/include/c++/8.2.0/cstdlib:132:11: error: ‘::aligned_alloc’ has not been declared
   using ::aligned_alloc;
           ^~~~~~~~~~~~~
gmake[3]: *** [CMakeFiles/pybind_test.dir/main.cpp.o] Error 1
gmake[2]: *** [CMakeFiles/pybind_test.dir/all] Error 2
gmake[1]: *** [CMakeFiles/pybind_test.dir/rule] Error 2
gmake: *** [pybind_test] Error 2
```

When I use `set(CMAKE_CXX_STANDARD 14)` or `set(CMAKE_CXX_STANDARD 11)` everything builds as expected. Interestingly, `gcc-7.3.0/bin/g++` works with any of `11`, `14` or `17`.

So the above error is exclusive to g++ `8.2.0` with `set(CMAKE_CXX_STANDARD 17)`. Anybody have an idea what the issue is?

---

Pybind11 version = 2.3.dev1 (taken from detail/common.h), as of 27 June 2019

**Comment by wjakob** [comment]
Your libstdc++ is potentially outdated and doesn't support all of C++17. (In particular, https://en.cppreference.com/w/c/memory/aligned_alloc)

**Comment by wjakob** [comment]
Whoops, I meant https://en.cppreference.com/w/cpp/memory/c/aligned_alloc.

**Comment by phil-zxx** [comment]
Since `libstdc++` is part of g++, I would expect 8.2 to be supporting more features than 7.3. So it's funny that 7.3 works, whereas 8.2 fails?

**Comment by phil-zxx** [comment]
It seemed to have been an issue with the **Python C++ Headers** (though I am unsure about the exact reason).

In case this helps someone else down the line, my fix was to downgrade the headers from **3.6.7** to **3.6.6**.
